src/misc/pretty_printing.ts

```diff
@@ -0,0 +1,71 @@
+import { ASTNode } from "../ast/ast_node";
+
+export interface PPAble {
+    pp(): string;
+}
+
+export type PPIsh = PPAble | ASTNode | string | number | boolean | bigint | null | undefined;
+
+export function isPPAble(value: any): value is PPAble {
+    return value ? typeof value.pp === "function" : false;
+}
+
+export function pp(value: PPIsh): string {
+    if (value instanceof ASTNode) {
+        return value.type + "#" + value.id;
+    }
+
+    if (value === undefined) {
+        return "<undefined>";
+    }
+
+    if (
+        value === null ||
+        typeof value === "string" ||
+        typeof value === "number" ||
+        typeof value === "boolean" ||
+        typeof value === "bigint"
+    ) {
+        return String(value);
+    }
+
+    if (isPPAble(value)) {
+        return value.pp();
+    }
+
```

Comment:
Could we add the following ifs here (to allow for pp() to handle (potentially nested) arrays, maps and iterables):

```
if (value instanceof Array) {
  return ppArr(value);
}

if (<value is an Iterable>) {
   return ppIter(value);
}

if (value instanceof Set) {
   return ppSet(value);
}

if (value instanceof Map) {
   return ppMap(value);
}
```

After we add this, we should add tests that we can pretty-print types like `string[][]`, `Map<string, Map<number, number>>`, `Set<string>[][]` etc...

Reply:
Addressed in https://github.com/ConsenSys/solc-typed-ast/commit/6e612113efada422eb324ed77841541855af8c10.

Still, we should be careful with possible recursion here.